File: 백준/Platinum/1786. 찾기/찾기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        char[] text = in.readLine().toCharArray();
        char[] pattern = in.readLine().toCharArray();

        int tLength = text.length, pLength = pattern.length;

        int[] pi = new int[pLength];
        for (int i = 1, j = 0; i < pLength; i++) {
            while (j > 0 && pattern[i] != pattern[j]) j = pi[j - 1];

            if (pattern[i] == pattern[j]) pi[i] = ++j;
            else pi[i] = 0;
        }

        int cnt = 0;
        ArrayList<Integer> list = new ArrayList<>();

        for (int i = 0, j = 0; i < tLength; ++i) {
            while (j > 0 && text[i] != pattern[j]) j = pi[j - 1];

            if (text[i] == pattern[j]) {
                if (j == pLength - 1) {
                    cnt++;
                    list.add(i - pLength + 2);
                    j = pi[j];
                } else {
                    j++;
                }
            }
        }

        System.out.println(cnt);
        if (cnt > 0) {
            for (Integer integer : list) {
                System.out.print(integer + " ");
            }
        }
    }
}

File: 백준/Platinum/1786. 찾기/찾기.py
import sys


def input(): return sys.stdin.readline().rstrip()


def make_pi(pattern):
    m = len(pattern)
    pi = [0] * m

    length = 0
    i = 1

    while i < m:
        if pattern[i] == pattern[length]:
            length += 1
            pi[i] = length
            i += 1
        else:
            if length != 0:
                length = pi[length - 1]
            else:
                pi[i] = 0
                i += 1

    return pi


def kmp_search(text, p):
    n = len(text)
    m = len(p)
    pi = make_pi(p)
    result = []

    i = 0
    j = 0

    while i < n:
        if p[j] == text[i]:
            i += 1
            j += 1

        if j == m:
            result.append(i - j + 1)
            j = pi[j - 1]

        elif i < n and p[j] != text[i]:
            if j != 0:
                j = pi[j - 1]
            else:
                i += 1

    return result


T = input()
P = input()

pos = kmp_search(T, P)

cnt = len(pos)
print(cnt)
print(*pos)